src/Microsoft.AspNetCore.Mvc.Core/Routing/DispatcherUrlHelper.cs

```diff
@@ -0,0 +1,160 @@
+        /// Initializes a new instance of the <see cref="UrlHelper"/> class using the specified
+        /// <paramref name="actionContext"/>.
+        /// </summary>
+        /// <param name="actionContext">The <see cref="Mvc.ActionContext"/> for the current request.</param>
+        /// <param name="linkGenerator">The <see cref="ILinkGenerator"/> used to generate the link.</param>
+        /// <param name="logger">The <see cref="ILogger"/>.</param>
+        public DispatcherUrlHelper(
+            ActionContext actionContext,
+            ILinkGenerator linkGenerator,
+            ILogger<DispatcherUrlHelper> logger)
+        {
+            if (actionContext == null)
+            {
+                throw new ArgumentNullException(nameof(actionContext));
+            }
+
+            if (linkGenerator == null)
+            {
+                throw new ArgumentNullException(nameof(linkGenerator));
+            }
+
+            ActionContext = actionContext;
+            _ambientValues = ActionContext.RouteData.Values;
+            _linkGenerator = linkGenerator;
+            _urlHelperCommon = new UrlHelperCommon(actionContext.HttpContext);
+            _logger = logger;
+        }
+
+        /// <inheritdoc />
+        public ActionContext ActionContext { get; }
+
+        /// <inheritdoc />
+        public virtual string Action(UrlActionContext actionContext)
+        {
+            if (actionContext == null)
+            {
+                throw new ArgumentNullException(nameof(actionContext));
+            }
+
+            var valuesDictionary = _urlHelperCommon.GetValuesDictionary(actionContext.Values);
+
+            if (actionContext.Action == null)
+            {
+                object action;
+                if (!valuesDictionary.ContainsKey("action") &&
+                    _ambientValues.TryGetValue("action", out action))
+                {
+                    valuesDictionary["action"] = action;
+                }
+            }
+            else
+            {
+                valuesDictionary["action"] = actionContext.Action;
+            }
+
+            if (actionContext.Controller == null)
+            {
+                object controller;
+                if (!valuesDictionary.ContainsKey("controller") &&
+                    _ambientValues.TryGetValue("controller", out controller))
+                {
+                    valuesDictionary["controller"] = controller;
+                }
+            }
+            else
+            {
+                valuesDictionary["controller"] = actionContext.Controller;
+            }
+
+            var successfullyGeneratedLink = _linkGenerator.TryGetLink(
+                new LinkGeneratorContext()
+                {
+                    SuppliedValues = valuesDictionary,
+                    AmbientValues = _ambientValues
+                },
+                out var link);
+
+            if (!successfullyGeneratedLink)
+            {
+                //TODO: log here
```

Comment:
or throw if some magic flag is set?

Reply:
Let's not try to design that experience in this PR

Reply:
Yeah, you are right. But I plan to address it with https://github.com/aspnet/Routing/issues/553